tests/server.py

```diff
@@ -917,13 +919,16 @@ def setup_test_homeserver(
     # Create the database before we actually try and connect to it, based off
     # the template database we generate in setupdb()
     if isinstance(db_engine, PostgresEngine):
+        import psycopg2.extensions
+
         db_conn = db_engine.module.connect(
             database=POSTGRES_BASE_DB,
             user=POSTGRES_USER,
             host=POSTGRES_HOST,
             port=POSTGRES_PORT,
             password=POSTGRES_PASSWORD,
         )
+        assert isinstance(db_conn, psycopg2.extensions.connection)
```

Comment:
A bit unfortunate we have to do this (and another instance below). I guess we could avoid it in the future by defining a `connect` on the engine which knows that it's returning a psycopg connection.

Reply:
(FTAOD I'm not suggesting we do anything about this now.)

Reply:
I think we *do* define the database engines as generic? I'll look again briefly.

Reply:
I suppose we could also make the `DBAPI2Module` generic over a connection type and plumb it through from the engine.

Reply:
I think fixing this is a bigger piece of work than I want to handle in this PR and would rather leave this as an exercise for a future reader?

Reply:
Fully agreed.